### ISSUE TITLE
[RFC] PHPMD Exception Interfaces

### ISSUE THREAD

**Issue by ravage84** [phpmd/phpmd#778]
Taken from @kylekatarnls review comment in https://github.com/phpmd/phpmd/pull/775

We could have interfaces for PHPMD exceptions:

- ``PHPMD\Exception\Exception``
- ``PHPMD\Exception\RuntimeException``
- ``PHPMD\Exception\InvalidArgumentException`` (just for the example)

and exceptions may implements them accordingly so the user is able to catch all ``PHPMD\Exception\Exception`` (instead of all \Exception) or all ``PHPMD\Exception\RuntimeException`` (instead of all ``\RuntimeException``).

**Comment by ravage84** [comment]
@kylekatarnls thanks for the proposal.

I agree, having a PHPMD library "base" exception (interface) like ``PHPMD\Exception\Exception`` would be a gread idea.

Also, it might make sense to have a ``PHPMD\Exception\LogicException`` and a ``PHPMD\Exception\RuntimeException`` to distinguish the kind of error.

But I'm not sure it would be much of a benefit to have any "deeper" exception double from the [SPL exception hierarchy](https://www.php.net/manual/en/spl.exceptions.php), like ``PHPMD\Exception\InvalidArgumentException``.

One additional interface might be useful:
``PHPMD\Exception\RuleException`` which is implemented by all rule related exceptions from PHPMD.

**Comment by MarkVaughn** [comment]
So the idea is to have `\PHPMD\Exception\RuntimeException` **extends** the `\RuntimeException` and **implements** `\PHPMD\Exception\Exception`?

**Comment by kylekatarnls** [comment]
Yes, this allows either to `catch(\RuntimeException $exception)` (all runtime exception from anywhere), or  `catch(\PHPMD\Exception\Exception $exception)` (any kind of exception from PHPMD) or  `catch(\PHPMD\Exception\RuntimeException $exception)` (the specific kind from PHPMD).

**Comment by tvbeek** [comment]
I like this idea 👍
